### PR TITLE
Contour

### PR DESCRIPTION
**Pull request by mattstoker**
Adds contour calculation for HCPath.
Adds drawing of contours in HCRaster.
Tests and memory fixes.

### CODE REVIEW

File: Source/Graphic/HCContourCurve.c [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+//    HCReal t2 = t * t;
+//    HCReal tc2 = tc * tc;
+//    HCReal x = p0.x * tc2 + (2.0 * c.x) * t * tc + p1.x * t2;
+//    HCReal y = p0.y * tc2 + (2.0 * c.y) * t * tc + p1.y * t2;
+    
+//    HCReal a = tc * tc;
+//    HCReal b = 2.0 * t * tc;
+//    HCReal c = t * t;
+//    HCReal x = a * p0.x + b * c.x + c * p1.x;
+//    HCReal y = a * p0.y + b * c.y + c * p1.y;
```
Comment (bscothern) [inline Source/Graphic/HCContourCurve.c]: Do you want all of these comments here?

File: Source/Graphic/HCContourCurve.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+//    HCReal tc3 = tc * tc * tc;
+//    HCReal x = p0.x * tc3 + (3.0 * c0.x) * t * tc2 + (3.0 * c1.x) * t2 * tc + p1.x * t3;
+//    HCReal y = p0.y * tc3 + (3.0 * c0.y) * t * tc2 + (3.0 * c1.y) * t2 * tc + p1.y * t3;
+    
+//    HCReal a = tc * tc * tc;
+//    HCReal b = 3.0 * t * tc * tc;
+//    HCReal c = 3.0 * t * t * tc;
+//    HCReal d = t * t * t;
+//    HCReal x = a * p0.x + b * c0.x + c * c1.x + d * p1.x;
+//    HCReal y = a * p0.y + b * c0.y + c * c1.y + d * p1.y;
```
Comment (bscothern) [inline Source/Graphic/HCContourCurve.c]: Do you want all these comments here?

File: Source/Graphic/HCContourCurve.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    HCPoint c0;
+    HCPoint c1;
+    HCPoint p;
+} HCContourCurve;
+
+//----------------------------------------------------------------------------------------------------------------------------------
+// MARK: - Contour Curve Special Values
+//----------------------------------------------------------------------------------------------------------------------------------
+#define HCContourCurveInvalidStatic {.c0 = HCPointInvalidStatic, .c1 = HCPointInvalidStatic, .p = HCPointInvalidStatic}
+#define HCContourCurveZeroStatic    {.c0 = HCPointInvalidStatic, .c1 = HCPointInvalidStatic, .p = HCPointZeroStatic}
```
Comment (bscothern) [inline Source/Graphic/HCContourCurve.h]: We probably want a type case before the `Static` literal values here.

File: Source/Graphic/HCPath+Shapes.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    HCPathMove(self, HCRectangleMinX(rectangle), HCRectangleMinY(rectangle));
+    HCPathAddLine(self, HCRectangleMaxX(rectangle), HCRectangleMinY(rectangle));
+    HCPathAddLine(self, HCRectangleMaxX(rectangle), HCRectangleMaxY(rectangle));
+    HCPathAddLine(self, HCRectangleMinX(rectangle), HCRectangleMaxY(rectangle));
+    HCPathClose(self);
+    return self;
+}
+
+HCPathRef HCPathCreateEllipse(HCRectangle ellipseBounds) {
+    static HCReal controlAdjustment = 0.552284749830793; //4.0 / 3.0 * (tan(M_PI * 0.5) * 0.25);
```
Comment (bscothern) [inline Source/Graphic/HCPath+Shapes.c:26]: 👍

File: Source/Graphic/HCPath.c [hunk 5]
```diff
@@ -230,21 +205,362 @@ HCPathElement HCPathElementAt(HCPathRef self, HCInteger elementIndex) {
...
+            }
+        }
+    }
+    return closedContourPaths;
+}
+
+HCPathRef HCPathOpenContoursAsPathRetained(HCPathRef self) {
+    HCListRef openContours = HCPathOpenContourPathsRetained(self);
+    HCPathRef openPath = HCPathCreateWithSubpaths(openContours);
+    HCRelease(openContours);
```
Comment (bscothern) [inline Source/Graphic/HCPath.c:377]: Does it make sense to have a `HCPathCreate` that takes ownership of the given contours?

File: Source/Graphic/HCPath.c [hunk 6]
```diff
@@ -230,21 +205,362 @@ HCPathElement HCPathElementAt(HCPathRef self, HCInteger elementIndex) {
...
 
+const HCPoint* HCPathPolylineAt(HCPathRef self, HCInteger polylineIndex) {
+    return (HCPoint*)HCDataBytes(HCListObjectAtIndex(self->polylines, polylineIndex));
+}
+
+//----------------------------------------------------------------------------------------------------------------------------------
+// MARK: - Polyline / Path Element Correspondence
+//----------------------------------------------------------------------------------------------------------------------------------
+HCInteger HCPathIndexOfPolylineContainingElement(HCPathRef self, HCInteger elementIndex) {
+    (void)self; // unused
```
Comment (bscothern) [inline Source/Graphic/HCPath.c:414]: Do we need this?

File: Source/Graphic/HCPath.c [hunk 7]
```diff
@@ -230,21 +205,362 @@ HCPathElement HCPathElementAt(HCPathRef self, HCInteger elementIndex) {
...
+        elementIndex++;
+    }
+    
+    // An odd intersection count indicates the point is within the path
+    return intersectionCount % 2 == 1;
+}
+
+HCBoolean HCPathContainsPointNonZero(HCPathRef self, HCPoint point) {\
+    (void)self; // unused
+    (void)point; // unused
```
Comment (bscothern) [inline Source/Graphic/HCPath.c:483]: Do we need these?

File: Source/Graphic/HCPath.c [hunk 8]
```diff
@@ -230,21 +205,362 @@ HCPathElement HCPathElementAt(HCPathRef self, HCInteger elementIndex) {
...
+            // Use the end point of the polyline line segment as the start point of the next
+            p0 = p1;
+        }
+    }
+}
+
+void HCPathIntersects(void* context, HCBoolean* stopSearching, HCPathRef path, HCPathRef otherPath, HCPoint point) {
+    (void)path; // unused
+    (void)otherPath; // unused
+    (void)point; // unused
```
Comment (bscothern) [inline Source/Graphic/HCPath.c:556]: Do we need these?

File: Source/Graphic/HCPath.c [hunk 9]
```diff
@@ -349,87 +658,161 @@ void HCPathAddElement(HCPathRef self, HCPathCommand command, const HCPoint* poin
...
     }
     self->bounds = bounds;
 }
 
 //----------------------------------------------------------------------------------------------------------------------------------
-// MARK: - Polylines
+// MARK: - Polyline Conversion
 //----------------------------------------------------------------------------------------------------------------------------------
 void HCPathAddLineSegmentPolylineData(HCPathRef self, HCReal x0, HCReal y0, HCReal x1, HCReal y1, HCDataRef polylineData) {
+    (void)self; // unused
```
Comment (bscothern) [inline Source/Graphic/HCPath.c:830]: Is this needed?

File: Source/Graphic/HCRaster+Draw.c [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+        HCReal a = tc;
+        HCReal b = t;
+        HCReal x = a * x0 + b * x1;
+        HCReal y = a * y0 + b * y1;
+        
+        HCRasterDrawPoint(self, x, y, HCColorCombine(c0, c1, t));
+        t += fmax(0.00001, 1.00000 / fmax(fabs(dx), fabs(dy)));
+    }
+    
+    // Draw using Bresenham's Algorithm
```
Comment (bscothern) [inline Source/Graphic/HCRaster+Draw.c]: Do we want these other algorithms to stay here?

File: Source/Graphic/HCRaster+Draw.c [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+//        for (HCReal x = xpxl1 + 1.0; x < xpxl2 - 1.0; x += 1.0) {
+//            HCRasterSetPixelAt(self, x, floor(intery),   HCColorCombine(HCColorBlack, c1, rfpart(intery)));
+//            HCRasterSetPixelAt(self, x, floor(intery)+1, HCColorCombine(HCColorBlack, c1,  fpart(intery)));
+//            intery = intery + gradient;
+//        }
+//    }
+}
+
+void HCRasterDrawQuadraticCurve(HCRasterRef self, HCReal x0, HCReal y0, HCReal cx, HCReal cy, HCReal x1, HCReal y1, HCColor c0, HCColor c1) {
+    // Draw using direct evaluation informed by derivitive
```
Comment (bscothern) [inline Source/Graphic/HCRaster+Draw.c]: Do we want this algorithm to stay here?

File: Source/Graphic/HCRaster+Draw.c [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+    HCReal qy1 =  cy * 0.5 +  y1 * 0.5;
+    HCReal  sx = qx0 * 0.5 + qx1 * 0.5;
+    HCReal  sy = qy0 * 0.5 + qy1 * 0.5;
+    HCColor cs = HCColorCombine(c0, c1, 0.5);
+    HCRasterDrawQuadraticCurve(self, x0, y0, qx0, qy0, sx, sy, c0, cs);
+    HCRasterDrawQuadraticCurve(self, sx, sy, qx1, qy1, x1, y1, cs, c1);
+}
+
+void HCRasterDrawCubicCurve(HCRasterRef self, HCReal x0, HCReal y0, HCReal cx0, HCReal cy0, HCReal cx1, HCReal cy1, HCReal x1, HCReal y1, HCColor c0, HCColor c1) {
+    // Draw using direct evaluation informed by derivitive
```
Comment (bscothern) [inline Source/Graphic/HCRaster+Draw.c]: Do we want this algorithm to stay here?

File: Source/Graphic/HCRaster+File.c [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+    }
+    
+    // Read pixel information
+    for (HCInteger yIndex = 0; yIndex < HCRasterHeight(self); yIndex++) {
+        for (HCInteger xIndex = 0; xIndex < HCRasterWidth(self); xIndex++) {
+            int r = 0;
+            int g = 0;
+            int b = 0;
+            if (!binary) {
+                fscanf(f, "%i %i %i", &r, &g, &b);
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c]: Should we ensure we are reading 3 values here? If it fails that means we have an invalid file and should probably fail.

File: Source/Graphic/HCRaster+File.c [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+            }
+            else {
+                fread(&r, 1, 1, f);
+                fread(&g, 1, 1, f);
+                fread(&b, 1, 1, f);
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c]: Should we also ensure we are reading all the values here?

File: Source/Graphic/HCRaster+File.c [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+            fprintf(file, "\n");
+        }
+    }
+    
+    // Cleanup
+    fclose(file);
+}
+
+HCRasterRef HCRasterCreateByLoadingBMP(const char* path) {
+    // BMP (Windows V3)
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c:148]: 👍

File: Source/Graphic/HCRaster+File.c [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+    int32_t width = 0;
+    int32_t height = 0;
+    //uint16_t colorPlanes = 1;
+    //uint16_t bitsPerPixel = 32;
+    //uint32_t compression = 0;
+    //uint32_t dataSize = width * height * bitsPerPixel / 8;
+    //int32_t horizontalResolution = 2835;
+    //int32_t verticalResolution = 2835;
+    //uint32_t paletteColorCount = 0;
+    //uint32_t importantPaletteColorCount = 0;
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c:190]: Do you want these here?

File: Source/Graphic/HCRaster+File.c [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    fread(&height, sizeof(int32_t), 1, f);
+    fseek(f, dataOffset, SEEK_SET);
+    //fread(&colorPlanes, sizeof(uint16_t), 1, f);
+    //fread(&bitsPerPixel, sizeof(uint16_t), 1, f);
+    //fread(&compression, sizeof(uint32_t), 1, f);
+    //fread(&dataSize, sizeof(uint32_t), 1, f);
+    //fread(&horizontalResolution, sizeof(uint32_t), 1, f);
+    //fread(&verticalResolution, sizeof(uint32_t), 1, f);
+    //fread(&paletteColorCount, sizeof(uint32_t), 1, f);
+    //fread(&importantPaletteColorCount, sizeof(uint32_t), 1, f);
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c:209]: Do you want these here?

File: Source/Graphic/HCRaster+File.c [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+    
+    return self;
+}
+
+void HCRasterSaveBMP(HCRasterRef self, const char* path) {
+    HCRasterSaveBMPWithOptions(self, path, false);
+}
+
+void HCRasterSaveBMPWithOptions(HCRasterRef self, const char* path, HCBoolean reversed) {
+    // BMP (Windows V3)
```
Comment (bscothern) [inline Source/Graphic/HCRaster+File.c:252]: 👍

**Review by bscothern — CHANGES_REQUESTED** [review]
Mostly some questions and small comments. I do think we should do the error handling though that I suggested in a couple places.

File: Source/Graphic/HCPath.c (hunk 5 above)
Comment (mattstoker) [reply Source/Graphic/HCPath.c:377]: It does, but construction with path elements of the same mechanism hasn't been implemented either. Nor efficient transformation of paths using a transform matrix. All of these should probably be done at the same time.

File: Source/Graphic/HCRaster+File.c (hunk 14 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+File.c:88]: The file format indicates that the image be of the size indicated in the header, read above. By having exactly the number of read calls as specified by the header, the implementation guarantees exactly that number of pixels result. Having to much data (or not enough) only needs to guarantee the same operation across systems, which this does by initializing r,g,b to 0.

File: Source/Graphic/HCRaster+File.c (hunk 16 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+File.c:190]: Yes, they are related to the TODO: on 206

File: Source/Graphic/HCRaster+File.c (hunk 17 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+File.c:209]: Yes, they are related to the TODO: on 206

File: Source/Graphic/HCContourCurve.h (hunk 3 above)
Comment (mattstoker) [reply Source/Graphic/HCContourCurve.h]: Done.

File: Source/Graphic/HCPath.c (hunk 6 above)
Comment (mattstoker) [reply Source/Graphic/HCPath.c:414]: Yes, otherwise compiling with -Wall gives a warning.

File: Source/Graphic/HCPath.c (hunk 7 above)
Comment (mattstoker) [reply Source/Graphic/HCPath.c:483]: Yes, otherwise compiling with -Wall gives a warning.

File: Source/Graphic/HCPath.c (hunk 8 above)
Comment (mattstoker) [reply Source/Graphic/HCPath.c:556]: Yes, otherwise compiling with -Wall gives a warning.

File: Source/Graphic/HCPath.c (hunk 9 above)
Comment (mattstoker) [reply Source/Graphic/HCPath.c:830]: Yes, otherwise compiling with -Wall gives a warning.

File: Source/Graphic/HCRaster+Draw.c (hunk 10 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+Draw.c]: No, but using Bresenham's with a color interpolating t value made things look MUCH nicer. Thanks for reminding me!

File: Source/Graphic/HCRaster+Draw.c (hunk 11 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+Draw.c]: Nope.

File: Source/Graphic/HCRaster+Draw.c (hunk 12 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+Draw.c]: Nope

File: Source/Graphic/HCRaster+File.c (hunk 13 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+File.c]: Now fails.

File: Source/Graphic/HCRaster+File.c (hunk 14 above)
Comment (mattstoker) [reply Source/Graphic/HCRaster+File.c]: Now fails.

**Review by bscothern — APPROVED** [review]

**Pull request merged**